ao-tools/ao-test-pressure/ao-test-pressure.1
.\"
.\" Copyright © 2022 Bdale Garbee <bdale@gag.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH AO-LOAD 1 "ao-test-pressure" ""
.SH NAME
ao-test-baro \- test AltOS motor testing pressure sensor
.SH SYNOPSIS
.B "ao-test-pressure"
[\-T \fItty-device\fP]
[\--tty \fItty-device\fP]
[\-D \fIaltos-device\fP]
[\--device \fIaltos-device\fP]
\fIpressure-name...\fP
.SH DESCRIPTION
.I ao-test-pressure
makes sure the pressure sensor input is near mid-band when driven by a 2:1
resistive divider across the sensor power rail.
.SH OPTIONS
.TP
\-T tty-device | --tty tty-device
This selects which tty device the debugger uses to communicate with
the target device.
.TP
\-D AltOS-device | --device AltOS-device
Search for a connected device. This requires an argument of one of the
following forms:
.IP
TeleMega:2
.br
TeleMega
.br
2
.IP
Leaving out the product name will cause the tool to select a suitable
product, leaving out the serial number will cause the tool to match
one of the available devices.
.SH USAGE
.I ao-test-pressure
opens the target device and queries the current pressure sensor input data
.SH AUTHOR
Bdale Garbee


ChangeLog
commit 8e42773e7834638464b46495123dbdc2356032bf
Merge: a6c32739 7fe1408a
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Jan 19 22:55:56 2023 -0700

    Merge branch 'master' into branch-1.9

commit 7fe1408af6fd65c13cebf33d7bb7bf8517d8ce1e
Merge: 234b275d 5f007934
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Jan 19 22:51:27 2023 -0700

    Merge branch 'master' of ssh://git.gag.com/scm/git/fw/altos

commit 5f007934c12863b5df1adc8d286b20b3488b5a6f
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:28:36 2023 -0800

    Version 1.9.13
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 97e6f06783006f59dd596d4f3d741408a1e4e5b1
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:27:18 2023 -0800

    doc: Update copyright year to 2023
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 1803c15b9d9110df9e54374b2576681835d6cd87
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:27:02 2023 -0800

    doc: Add release notes for 1.9.13
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 96181ff1e0d376d7a2c9519abbf4f346d8fd47ea
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Jan 19 22:41:24 2023 -0700

    doc: update EasyMotor behavior description for firmware 1.9.13 and later

commit 234b275db125304829d1859713388775fd3791db
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Jan 19 22:41:24 2023 -0700

    doc: update EasyMotor behavior description for firmware 1.9.13 and later

commit 22f9fb47d97ba13f6b30fab25bce552bccc91997
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:39:42 2023 -0800

    Disable TM v4.0 firmware for 1.9.13
    
    This will be in 1.9.14
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 3021764629a438d1d1b0323a09d6fb2350ce1bfb
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:28:36 2023 -0800

    Version 1.9.13
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 2689b30cdccba18d46505774aafc8e6119948191
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:27:18 2023 -0800

    doc: Update copyright year to 2023
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 399c85245542d95f05063fdd34939dc746f1ea2d
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:27:02 2023 -0800

    doc: Add release notes for 1.9.13
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 691869aa6c542999a4e915c23b6696451dd42f65
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 17:20:44 2023 -0800

    altos: Add makefiles for telemega v6
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit e4ce0ce186b68497cfb14a400410fdd38aa93abc
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 14:04:50 2023 -0800

    altos/easymotor: Move pressure conversion code to ao_motor_flight.c
    
    This inline function uses a pile of constants which aren't defined in
    ao_pins.h, so move it to ao_motor_flight.c where it is used
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 28e4cb24e0f0ee5abf66c5a0466edfd0e31f4df0
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 13:20:56 2023 -0800

    easymotor-v3: Fix pressure to adc conversion to use reference voltage
    
    Wasn't taking reference voltage into account, which meant the computed values
    were 3.3 times too high.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 7b40813fdafce4ebd635aceb4c32355339a4d98d
Merge: 026ab589 4c61c113
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 12:21:32 2023 -0800

    Merge branch 'easymotor-flight'

commit 026ab589d0ade3dd59d26888fffd8699ef534671
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Jan 19 12:18:46 2023 -0800

    altos/telelco: Add some debug aids for TeleLCO event stuff
    
    Use (debug & 2) to select just event debugging.
    Add 'pretend' mode to cycle through all possible pad boxes.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit ee49d3c202e1fee14328006e2beb2e774313de2e
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 27 21:33:14 2022 -0800

    altoslib: Add TeleMega v6 support
    
    New sensor (bmi088) support, plus other changes as needed. Also found
    some places missing TeleMega-v5 support?
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 96444c1e0e6a1c179d9695b454012d0b16d0a76c
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 27 21:32:39 2022 -0800

    ao-tools: Update ao-eeprom for TeleMega v6
    
    Add new log file format; no other changes needed here.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 95d39febfa9b6c6f2696fb1b47cf72aa808ec7c9
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 27 21:31:38 2022 -0800

    ao-tools: Fix buffer overflow in lib/ao-hex.c
    
    ao_hex_file_create allocated a buffer too short when writing EOF.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 08f13d3301bfcf9a5b9b566df4ffd4ed33f236d0
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 27 21:30:13 2022 -0800

    altos: Split AO_LOG_NORMALIZED support out in telemetry code
    
    This avoids needing per-driver changes to the telemetry code by using
    the normalized data gathering techniques.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 0d16160ce2207fd248c2d93a4f7a77a05176204d
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 27 21:29:12 2022 -0800

    drivers: Add AO_LOG_NORMALIZED support to BMI088 command
    
    Make this driver work in monitor idle mode
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit ad4679dd0bd2d759f708a6ab6d0425eb1cbde7c1
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 27 21:01:06 2022 -0800

    src/drivers: Ignore printf warnings in ao_aprs.c
    
    This code uses sprintf to generate the tnc buffer, but the compiler doesn't
    really understand the value limits.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>


ao-tools/ao-test-pressure/ao-test-pressure.c
/*
 * Copyright © 2022 Bdale Garbee <bdale@gag.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

#include <err.h>
#include <fcntl.h>
#include <gelf.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <sysexits.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <stdbool.h>
#include "ao-elf.h"
#include "ccdbg.h"
#include "cc-usb.h"
#include "cc.h"
#include "ao-verbose.h"

static const struct option options[] = {
	{ .name = "tty", .has_arg = 1, .val = 'T' },
	{ .name = "device", .has_arg = 1, .val = 'D' },
	{ .name = "raw", .has_arg = 0, .val = 'r' },
	{ .name = "verbose", .has_arg = 1, .val = 'v' },
	{ 0, 0, 0, 0},
};

static void usage(char *program)
{
	fprintf(stderr, "usage: %s [--verbose=<verbose>] [--device=<device>] [-tty=<tty>] main|drogue\n", program);
	exit(1);
}

static void
done(struct cc_usb *cc, int code)
{
/*	cc_usb_printf(cc, "a\n"); */
	cc_usb_close(cc);
	exit (code);
}

static char **
tok(char *line) {
	char	**strs = malloc (sizeof (char *)), *str;
	int	n = 0;

	while ((str = strtok(line, " \t"))) {
		line = NULL;
		strs = realloc(strs, (n + 2) * sizeof (char *));
		strs[n] = strdup(str);
		n++;
	}
	strs[n] = '\0';
	return strs;
}

static void
free_strs(char **strs) {
	char	*str;
	int	i;

	for (i = 0; (str = strs[i]) != NULL; i++)
		free(str);
	free(strs);
}

struct pressure {
	struct pressure	*next;
	char		**strs;
};

static struct pressure *
pressure(struct cc_usb *usb)
{
	struct pressure	*head = NULL, **tail = &head;
	cc_usb_printf(usb, "a\n");
	for (;;) {
		char	line[512];
		struct pressure	*b;

		cc_usb_getline(usb, line, sizeof (line));
		b = malloc (sizeof (struct pressure));
		b->strs = tok(line);
		b->next = NULL;
		*tail = b;
		tail = &b->next;
		if (strstr(line, "tick:"))
			break;
	}
	return head;
}

static void
free_pressure(struct pressure *b) {
	struct pressure *n;

	while (b) {
		n = b->next;
		free_strs(b->strs);
		free(b);
		b = n;
	}
}

static int
do_pressure(struct cc_usb *usb) {
	struct pressure *b = pressure(usb);
	char **alt = b->strs;

	if (!alt) {
		printf("no response\n");
		free_pressure(b);
		return 0;
	}

	double pressure = strtod(alt[5], NULL);

	if (pressure < 15000 || pressure > 16000) {
		printf ("weird pressure %f\n", pressure);
		free_pressure(b);
		return 0;
	}

	printf ("pressure %f\n", pressure);

	return 1;
}

int
main (int argc, char **argv)
{
	char			*device = NULL;
	int			c;
	struct cc_usb		*cc = NULL;
	char			*tty = NULL;
	int			verbose = 0;
	int			ret = 0;

	while ((c = getopt_long(argc, argv, "rT:D:c:s:v:", options, NULL)) != -1) {
		switch (c) {
		case 'T':
			tty = optarg;
			break;
		case 'D':
			device = optarg;
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage(argv[0]);
			break;
		}
	}

	ao_verbose = verbose;

	if (verbose > 1)
		ccdbg_add_debug(CC_DEBUG_BITBANG);

	if (!tty)
		tty = cc_usbdevs_find_by_arg(device, "AltosFlash");
	if (!tty)
		tty = cc_usbdevs_find_by_arg(device, "EasyMotor");
	if (!tty)
		tty = getenv("ALTOS_TTY");
	if (!tty)
		tty="/dev/ttyACM0";

	cc = cc_usb_open(tty);

	if (!cc)
		exit(1);

	if (!do_pressure(cc))
		ret = 1;
	done(cc, ret);
}
